c7n/resources/elasticsearch.py

```diff
@@ -303,6 +326,111 @@ def source_ips_from_stmt(cls, stmt):
         return source_ips
 
 
+@ElasticSearchDomain.filter_registry.register('upgrade-available')
+class UpgradeAvailable(Filter):
+    """Scans for available upgrade-compatible ES versions
+
+    This will check all the ElasticSearch domains on the resources, and return
+    a list of viable upgrade options.
+
+    :example:
+
+    .. code-block:: yaml
+
+            policies:
+              - name: elasticsearch-upgrade-available
+                resource: elasticsearch
+                filters:
+                  - type: upgrade-available
+                    major: False
+
+    """
+
+    schema = type_schema(
+        'upgrade-available',
+        major={'type': 'boolean'},
+        value={'type': 'boolean'},
+    )
+    permissions = ('es:GetCompatibleVersions',)
+
+    def process(self, resources, event=None):
+        client = local_session(self.manager.session_factory).client('es')
+        check_major = self.data.get('major', False)
+        check_upgrade_extant = self.data.get('value', True)
+        results = []
+
+        for r in resources:
+            domain_name = r["DomainName"]
+            current_version = r.get('ElasticsearchVersion', '')
+
+            # Get compatible versions for this domain
+            try:
+                response = self.manager.retry(
+                    client.get_compatible_elasticsearch_versions,
```

Comment:
its somewhat interesting this api exists, versus simply being based on a mapping of extant version to upgradable. I wonder if it is version based, in which case we could cache based on extant version, or if it's on the basis of feature/configuration in use.

Reply:
It is strange, especially because it maps to a differently named API call (https://docs.aws.amazon.com/opensearch-service/latest/APIReference/API_GetCompatibleVersions.html), which returns a **blend** of engines (not just ElasticSearch, but `OpenSearch` as well). My guess is that it's custom-patching over old API that people depended on, or similar.